### PR TITLE
RFC: Implement coinductive predicates #1621

### PR DESCRIPTION
**Pull request by johoelzl**
This implements the construction of coinductive predicates. Currently only a simple form is supported, i.e. no mutual or nested corecursion.

I'm not sure what to do with the `corec` rule, it currently shows directly the internal inductive predicate.

### CODE REVIEW

**Comment by leodemoura** [comment]
@johoelzl This is great. Are you going to try to reuse the C++ inductive parser?

**Comment by johoelzl** [comment]
Yes I want to reuse the inductive parser.

Currently I'm working on mutual & nested definitions. When this works I will take a look into the parser.

**Comment by johoelzl** [comment]
@leodemoura I added support for mutual coinductive predicates and the expected coinductive syntax. I'm not 100% sure if `shared_coinductive_cmd` is correctly implemented:
* do I need to create a `vm_state`, even when using `tactic_evaluator`? When not I get an assertion error that `g_vm_state` is not set.
* The `tactic_evaluator` either returns success or transforms the error into an c++ exception. Only when a silent exception happens nothing is done. What are the cases for a silent exception? 

What I want to implement in the next days:
- [x] syntax!
- [x] tests!
- [X]  nested recursion
- [X]  coinduction rule
- [X]  coinduction method
- [X]  producing a case rule
- [X]  setting up attributes

**Comment by jroesch** [comment]
@johoelzl You probably need to scope a VM state, things which require a global VM must have one scoped by using the class `scope_vm_state`. For example:

```
src/library/inductive_compiler/nested.cpp
2255:        scope_vm_state vms_scope(vms);
```

**Comment by johoelzl** [comment]
@jroesch I currently use `scope_vm_state`. But I was not sure if this is the right approach. My usage looks like the one in `nested.cpp`. So I guess I'm fine. Thanks.

**Comment by leodemoura** [comment]
@johoelzl I can take a look tonight. Please post here any information you think is relevant.
I need one example. Should I use `tests/lean/run/coinductive.lean`?

**Comment by johoelzl** [comment]
@leodemoura `coinduction` and `mutual coinduction` both work. I used the examples I added to `tests/lean/run/coinductive.lean`. My question w.r.t `scope_vm_state` is mostly if this is the right approach or not.

I didn't try any error cases yet. There are no error messages yet.

Essentially every predicate one can write down with `inductive` (without nested induction) should work now with `coinductive`.

**Comment by johoelzl** [comment]
Sorry accidentally closed the PR.

File: src/frontends/lean/inductive_cmds.cpp [hunk 1]
```diff
@@ -693,6 +697,61 @@ class inductive_cmd_fn {
...
+        vm_obj r = tactic_evaluator(m_ctx, opts, mk_true())
+            (mk_constant(get_tactic_add_coinductive_predicate_name()), args, s);
+        if (auto new_s = tactic::is_success(r)) {
+            m_env = new_s->env();
+            post_process(new_params, new_inds, new_intro_rules);
+            return m_env;
+        }
+
+        throw generic_exception(first_ind, "silent exception"); // TODO(): what should we report here?
+    }
```
Comment (leodemoura) [inline src/frontends/lean/inductive_cmds.cpp]: The `tactic_evaluator` already reported the tactic failure.
We can simply say that the `coinductive` command failed. Example:
```cpp
    throw generic_exception(first_ind, "coinductive command failed");
```

File: src/frontends/lean/inductive_cmds.cpp [hunk 2]
```diff
@@ -693,6 +697,61 @@ class inductive_cmd_fn {
...
+        buffer<vm_obj> inds_arg;
+        for (unsigned i = 0; i < new_inds.size(); i ++) {
+            inds_arg.push_back(mk_vm_pair(to_obj(new_inds[i]), to_obj(new_intro_rules[i])));
+        }
+        args.push_back(to_obj(m_lp_names));
+        args.push_back(to_obj(new_params));
+        args.push_back(to_obj(inds_arg));
+
+        vm_obj r = tactic_evaluator(m_ctx, opts, mk_true())
+            (mk_constant(get_tactic_add_coinductive_predicate_name()), args, s);
```
Comment (leodemoura) [inline src/frontends/lean/inductive_cmds.cpp]: The third argument of `tactic_evaluator` is used to provide position information for error messages.
`mk_true` does not have position information associated with it.

File: src/frontends/lean/inductive_cmds.cpp [hunk 3]
```diff
@@ -693,6 +697,61 @@ class inductive_cmd_fn {
...
+        }
+        elaborate_inductive_decls(params, inds, intro_rules, new_params, new_inds, new_intro_rules);
+
+        expr first_ind = inds[0];
+        metavar_context mctx;
+        options opts = m_p.get_options();
+
+        vm_state vms(m_env, opts);
+        scope_vm_state vms_scope(vms);
+        tactic_state s = mk_tactic_state_for(m_env, opts, mlocal_name(first_ind), mctx, local_context(), mk_true());
```
Comment (leodemoura) [inline src/frontends/lean/inductive_cmds.cpp]: I just pushed a relevant fix: 
https://github.com/leanprover/lean/commit/544817cf150a726dbc531e7eb816f1921c478868
The lines 
```lean
  vm_state vms(m_env, opts);
  scope_vm_state vms_scope(vms);
```
are not needed anymore. The `tactic_evaluator` creates its own VM.

File: library/init/meta/tactic.lean [hunk 4]
```diff
@@ -997,6 +997,20 @@ match _root_.try_for max (tac s) with
...
 end
 
+meta def updateex_env (f : environment → exceptional environment) : tactic unit :=
+do env ← get_env,
+   env ← returnex $ f env,
+   set_env env
+
+meta def add_decl (d : declaration) : tactic unit :=
+updateex_env $ λe, do e ← e.add d, e.compile d tt
+
```
Comment (leodemoura) [inline library/init/meta/tactic.lean]: What is the motivation for this change?
This seems to be a refactoring that has nothing to do with the coinductive command.
We need to discuss this kind of change. For example, I'm not sure I want to expose `compile` to users.

File: library/init/meta/coinductive_predicates.lean [hunk 5]
```diff
@@ -0,0 +1,326 @@
...
+| (mk_numeral _ n) := last_string n
+
+end name
+
+namespace expr
+open expr
+
+meta def replace_with (e : expr) (s : expr) (s' : expr) : expr :=
+e.replace $ λc d, if c = s then some s' else none
+
```
Comment (leodemoura) [inline library/init/meta/coinductive_predicates.lean]: `replace_with` will produce incorrect results if `s'` contains dangling de-Bruijn variables.
This is not an issue in your code since this case never occurs.
However, if the procedure `replace_with` is available, other users may invoke it with terms containing dangling deBruijn variables. I think we have two possible solutions: mark `replace_with` private, or use `lift_vars`.
```lean
meta def replace_with (e : expr) (s : expr) (s' : expr) : expr :=
e.replace $ λc d, if c = s then some (s'.lift_vars 0 d) else none
```

File: src/frontends/lean/inductive_cmds.cpp (hunk 2 above)
Comment (johoelzl) [reply src/frontends/lean/inductive_cmds.cpp]: Then I will use `first_ind` to return errors. I guess this will report the error at the first identifier. Or is it possible to encode the beginning of the `coinductive` command as `expr `?

File: src/frontends/lean/inductive_cmds.cpp (hunk 3 above)
Comment (johoelzl) [reply src/frontends/lean/inductive_cmds.cpp]: Nice. I will remove them.

File: library/init/meta/tactic.lean (hunk 4 above)
Comment (johoelzl) [reply library/init/meta/tactic.lean]: It felt cleaner to provide more fine grained operations to the Lean side. I will revert `add_decl`, to be implemented in C++ again.

File: library/init/meta/coinductive_predicates.lean (hunk 5 above)
Comment (johoelzl) [reply library/init/meta/coinductive_predicates.lean]: Okay, I will add `lift_vars`.

File: src/frontends/lean/inductive_cmds.cpp (hunk 2 above)
Comment (leodemoura) [reply src/frontends/lean/inductive_cmds.cpp]: The `first_ind` is good. We use the approach in many other commands.

File: tests/lean/run/coinductive.lean [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+
+#print prefix tt_stream
+#print prefix ff_stream
+
+#check (@tt_stream : stream bool → Prop)
+#check (@tt_stream.corec_functional :
+  ∀ (C_tt_stream C_ff_stream : stream bool → Prop),
+    (∀ (a : stream bool), C_tt_stream a → tt_stream.functional C_tt_stream C_ff_stream a) →
+    (∀ (a : stream bool), C_ff_stream a → ff_stream.functional C_tt_stream C_ff_stream a) →
+    ∀ (a : stream bool), C_tt_stream a → tt_stream a)
```
Comment (leodemoura) [inline tests/lean/run/coinductive.lean:51]: Are you planning to hide the auxiliary type `*.functional`?

Comment (johoelzl) [reply tests/lean/run/coinductive.lean:51]: `*.functional` is not mentioned in the constructor rules or `corec_on` and `cases_on` anymore. I can put anything having `*.functional` in its type into the `*.functional` namespace. I'm not sure about hiding it, sometimes the construction is interesting in itself.

Is there an easy way to hide the definitions?

**Comment by johoelzl** [comment]
All features are implemented. I will add some tests for the `coinduction` tactic.

A couple of infrastructure changes are still missing:
* add an attribute to register coinductive rules, so the user does not need to write down the name of the rule
* add support for `cases` and `case`

**Comment by leodemoura** [comment]
@johoelzl I think we can merge it as is, right?
The missing features can be in different PRs.

**Comment by johoelzl** [comment]
Yes we can merge it.
